manim/mobject/types/vectorized_mobject.py

```diff
@@ -2117,7 +2127,44 @@ def construct(self):
                         (gr-circle_red).animate.shift(RIGHT)
                     )
         """
-        return super().add(*vmobjects)
+
+        def get_type_error_message(invalid_obj, invalid_i):
+            return (
+                f"Only values of type {vmobject_render_type.__name__} can be added "
+                f"as submobjects of VGroup, but the value "
+                f"{repr(invalid_obj)} (at index {invalid_i}) is of type "
+                f"{type(invalid_obj).__name__}."
+            )
+
+        vmobject_render_type = (
+            OpenGLVMobject if config.renderer == RendererType.OPENGL else VMobject
+        )
+        valid_vmobjects = []
+
+        for vmobject_i, vmobject in enumerate(vmobjects):
+            if isinstance(vmobject, vmobject_render_type):
+                valid_vmobjects.append(vmobject)
+            elif isinstance(vmobject, Iterable) and not isinstance(
+                vmobject, (Mobject, OpenGLMobject)
+            ):
+                for subvmobject_i, subvmobject in enumerate(vmobject):
+                    if not isinstance(subvmobject, vmobject_render_type):
+                        raise TypeError(
+                            get_type_error_message(subvmobject, subvmobject_i)
```

Comment:
Maybe something like this would be more helpful to the user?
```suggestion
                            get_type_error_message(subvmobject, (vmobject_i, subvmobject_i))
```
As a side note, maybe variable names like `i` for `vmobject_i` and `j` for `subvmobject_i` might be a little bit less verbose?

Reply:
I was also thinking something like this, but adding the `vmobject_i` to these `TypeError` messages will break some of the tests as they are expecting error messages with only a single index. What do you think I should do? 🙂

Reply:
Also I will simplify the variable names like you suggested 👍

Reply:
> Thanks for the feedback! I was also thinking something like this, but adding the `vmobject_i` to these `TypeError` messages will break some of the existing tests as they are expecting error messages with only a single index. What do you think I should do? 🙂 

I think in this case it's a good idea to change the tests. I would rather have more information for debugging than less!

Reply:
Good point! I've updated the tests and updated the error message to be more descriptive like you suggested. [Here's the updated error message in the code](https://github.com/ManimCommunity/manim/pull/3966/files#diff-3f967d53f766c19d36e76e161c5d1ac70e25e4f789415cd92615424981225f22R2149-R2156), let me know if you need me to make it more descriptive!